### DISCUSSION TITLE
Feature Request: make substitution return a new term instead of changing an existing term

### DISCUSSION THREAD

**Discussion by baierd** [bitwuzla/bitwuzla#87] [Ideas]
Hi,

as far as i understand the substitution implementation, it takes an existing term and mutates it.
This is dangerous, as usages of the term might be affected without noticing.
I propose to return a new term, that is a copy of the entered term, besides the substitution.

An example use-case would be quantifiers. If we create a formula and want to use it inside and outside of a quantified term, i would need to replace the free variables with bound ones. This would however mutate the formula outside of the quantifier as well.
(This is the exact use-case that we need substitution for in JavaSMT :D )

Would such a function be possible?

Best,
Daniel

**Comment by mpreiner** [comment]
Terms are immutable and will never be modified. The substitution does return new terms.

**Comment by baierd** [comment]
OH. I just assumed that they don't because of the void return type for mutliple terms. My bad. Sorry.
